proof/refine/ARM/SchedContext_R.thy

```diff
@@ -799,4 +799,74 @@ lemmas getSchedContext_setSchedContext_decompose_decompose2
 lemmas getSchedContext_setSchedContext_decompose_decompose_ext2
   = getSchedContext_setSchedContext_decompose[where f="f x" and g="g y" for f g x y]
 
+(* rewrite rules for updateSchedCOntext *)
+lemma updateSchedContext_decompose:
+   "monadic_rewrite False True
+     (sc_at' scPtr and K (\<forall>sc. objBits (f sc) = objBits sc) and K (\<forall>sc. objBits (g sc) = objBits sc))
+     (updateSchedContext scPtr (g o f))
+     (do updateSchedContext scPtr f;
+         updateSchedContext scPtr g
+      od)"
+  unfolding updateSchedContext_def bind_assoc o_def
+  using getSchedContext_setSchedContext_decompose by blast
+
+lemma updateSchedContext_decompose_twice:
+  "\<lbrakk>\<forall>sc. objBits (f sc) = objBits sc; \<forall>sc. objBits (g sc) = objBits sc;
+    \<forall>sc. objBits (h sc) = objBits sc\<rbrakk> \<Longrightarrow>
+    monadic_rewrite False True
+     (sc_at' scPtr)
+     (updateSchedContext scPtr (h o g o f))
+     (do updateSchedContext scPtr f;
+         updateSchedContext scPtr g;
+         updateSchedContext scPtr h
+      od)"
+  apply (rule monadic_rewrite_imp)
+   apply (rule monadic_rewrite_trans)
+    apply (rule updateSchedContext_decompose)
+   apply (rule monadic_rewrite_bind_tail)
+    apply simp
+    apply (rule updateSchedContext_decompose[simplified])
+   apply (wpsimp wp: updateSchedContext_wp)
+  apply (clarsimp simp: obj_at_simps opt_map_def ps_clear_upd)
+  done
+
+lemma updateSchedContext_decompose_thrice:
+  "\<lbrakk>\<forall>sc. objBits (f sc) = objBits sc; \<forall>sc. objBits (g sc) = objBits sc;
+    \<forall>sc. objBits (h sc) = objBits sc;  \<forall>sc. objBits (k sc) = objBits sc\<rbrakk> \<Longrightarrow>
+    monadic_rewrite False True
+     (sc_at' scPtr)
+     (updateSchedContext scPtr (k o h o g o f))
+     (do updateSchedContext scPtr f;
+         updateSchedContext scPtr g;
+         updateSchedContext scPtr h;
+         updateSchedContext scPtr k
+      od)"
```

Comment:
What's the context for using this rule?

I'm *guessing* that it would be where you have a single update of the form:
```
updateSchedContext scPtr (%sc. sc(| foo := k (foo sc), bar := h (bar sc), etc |))
```
in which case I guess you'd first need to do an ordinary (non-monadic) rewrite to the form:
```
updateSchedContext scPtr (foo_update (K k) o bar_update (K h) o ...)
```
before you can do this `monadic_rewrite`.

If so, then I think it would be nicer to have just a *single* `updateSchedContext_decompose` rule that rewrites a fold into a `mapM_x`, proved by induction:
```
lemma updateSchedContext_decompose:
  "\<forall>sc f. f : set fs --> objBits (f sc) = objBits sc \<Longrightarrow>
   monadic_rewrite False True
     (sc_at' scPtr)
     (updateSchedContext scPtr (fold (o) fs))
     (mapM_x (updateSchedContext scPtr) fs)"
```

Then for the first non-monadic rewrite above, you would instead rewrite to:
```
updateSchedContext scPtr (fold (o) [foo_update (K k), bar_update (K h), ...])
```

Another way to get exactly the `monadic_rewrite` you need from this list-based `updateSchedContext_decompose` would be to use:
```
updateSchedContext_deompose[where fs="[update_foo (K k), update_bar (K h), ...]", simplified]
```
With different length lists, the `fold` and `mapM_x` would expand out to give essentially your three rules, and the composition  of `update_foo` etc. would collapse into the `sc(| ... |)` form.

I haven't attempted to type-check any of this, so I probably got a bunch of details wrong. For example, I might not have the right `fold`, and if you pick the wrong fold, you might need  to `reverse fs` in the `mapM_x`.

Reply:
Thanks, Matt. These are used in #290 , and the rewrite is done in the opposite direction. This is to bundle up multiple updates on the concrete side to match with a single step update on the abstract side. So, simplification from the record notation to composed updates that you mentioned is not needed there.

But I do agree that these rewrite rules could do with a bit more sophistication, probably along the line of
```
updateSchedContext_deompose[where fs="[update_foo (K k), update_bar (K h), ...]", simplified]
```
I think Michael had something similar, but that was mixed with function extensionality. But I believe there is something in common with what you say here that I can use. What I might do is to move lemmas to #290 and tweak them more there.

Reply:
Right, that makes sense. You might still be able to get a modest improvement by proving a `monadic_rewrite` rule with `fold (o)` and `mapM_x`, and getting your three rules with something like:
```
lemmas updateSchedContext_decompose
  = updateSchedContext_decompose_fold[where fs="[f, g]" for f g, simplified]
lemmas updateSchedContext_decompose_twice
  = updateSchedContext_decompose_fold[where fs="[f, g, h]" for f g h, simplified]
lemmas updateSchedContext_decompose_thrice
  = updateSchedContext_decompose_fold[where fs="[f, g, h, k]" for f g h k, simplified]
```